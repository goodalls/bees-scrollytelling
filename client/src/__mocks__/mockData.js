module.exports = {
  appDefaultState: {
    scrollPosition: 750,
    questionsEnabled: true,
    currentUserId: null,
    displayGraphCover: true
  },
  event: {
    target: {
      name: 'Jeff',
      value: 'websocket'
    }
  },
  beeFactsPost: {
    body: '{"users_id":2,"user_answer":"","question":"bee species"}',
    headers: { 'Content-type': 'application/json' },
    method: 'POST'
  },
  headerDefaultState: {
    concern: '',
    age: '',
    location: {},
    answered: false
  },
  impactDefaultState: {
    percent: 0,
    unit: 'million',
    monetary: '',
    economy: '',
    answeredeconomy: false,
    answeredpercent: false
  },
  impactFactsPost: {
    body: '{"users_id":2,"question":"are plants great"}',
    headers: { 'Content-type': 'application/json' },
    method: 'POST'
  },
  problemsPostBody: {
    method: 'POST',
    headers: {
      'Content-type': 'application/json'
    },
    body: JSON.stringify({
      users_id: 2,
      user_answer: 30,
      question: 'What percent of bee colonies do beekeepers lose every year?'
    })
  },
  rawStatistics: [
    {
      id: 48,
      question:
        'What percentage of your diet do you think is provided by bees?',
      user_answer: '35',
      users_id: 50
    },
    {
      id: 49,
      question: 'How much do bees contribute to the economy, anually?',
      user_answer: '400_million',
      users_id: 50
    },
    { id: 50, question: 'bee species', user_answer: '200', users_id: 50 },
    {
      id: 51,
      question: 'What percent of bee colonies do beekeepers lose every year?',
      user_answer: '40',
      users_id: 50
    },
    {
      id: 52,
      question:
        'What percentage of your diet do you think is provided by bees?',
      user_answer: '80',
      users_id: 51
    },
    {
      id: 53,
      question: 'How much do bees contribute to the economy, anually?',
      user_answer: '400_billion',
      users_id: 51
    },
    { id: 54, question: 'bee species', user_answer: '100', users_id: 51 },
    {
      id: 55,
      question: 'What percent of bee colonies do beekeepers lose every year?',
      user_answer: '25',
      users_id: 51
    }
  ],
  cleanedStatistics: {
    'How much do bees contribute to the economy, anually?': {
      count: { '400_million': 1, '400_billion': 1 },
      sum: 400400000000
    },
    'What percent of bee colonies do beekeepers lose every year?': {
      count: { 25: 1, 40: 1 },
      sum: 65
    },
    'What percentage of your diet do you think is provided by bees?': {
      count: { 35: 1, 80: 1 },
      sum: 115
    },
    'bee species': { count: { 100: 1, 200: 1 }, sum: 300 }
  },
  cleanedAverageData: {
    'How much do bees contribute to the economy, anually?': {
      count: { '400_million': 1, '400_billion': 1 },
      sum: 400400000000,
      average: 200200000000
    },
    'What percent of bee colonies do beekeepers lose every year?': {
      count: { 25: 1, 40: 1 },
      sum: 65,
      average: 32.5
    },
    'What percentage of your diet do you think is provided by bees?': {
      count: { 35: 1, 80: 1 },
      sum: 115,
      average: 57.5
    },
    'bee species': { count: { 100: 1, 200: 1 }, sum: 300, average: 150 }
  },
  pieChartDefaultState: {
    class: '',
    degree: 0,
    percent: 10
  },
  percentEvent: {
    target: {
      name: 'percent',
      value: 55
    }
  }
};
