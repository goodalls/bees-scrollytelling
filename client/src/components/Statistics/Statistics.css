.Statistics {
  display: flex;
  flex-direction: column;
  align-items: center; }
  .Statistics h2 {
    font-size: 3em; }
  .Statistics h4, .Statistics h6 {
    font-size: 1.5em; }

.statistics-card {
  display: flex;
  flex-direction: column;
  border: 1px solid #e0e0e0;
  width: 100%;
  align-items: center;
  box-shadow: 5px 5px 5px #e0e0e0;
  margin: 10px; }

.answer {
  border: 1px solid magenta; }

hr {
  border: 1px solid #60CA9B;
  width: 50%; }
