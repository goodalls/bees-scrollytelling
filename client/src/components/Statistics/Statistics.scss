@import '../../assets/variables.scss';

.Statistics {
  display: flex;
  flex-direction: column;
  align-items: center;
  

  & h2 {
    font-size: 3em;
  }

  & h4, h6 {
    font-size: 1.5em;
  }
}

.statistics-card {
  display: flex;
  flex-direction: column;
  border: 1px solid $shadowgray;
  width: 100%;
  align-items: center;
  box-shadow: 5px 5px 5px $shadowgray;
  margin: 10px;
}

.answer {
  border: 1px solid magenta;
}

hr {
  border: 1px solid $seafoam;
  width: 50%;
}
