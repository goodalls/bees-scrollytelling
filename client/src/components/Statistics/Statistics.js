import React, { Component } from 'react';
import './Statistics.css';

class Statistics extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  componentDidMount = async () => {
    await this.getStatistics();
  }

  getStatistics = async () => {
    const output = await fetch('/api/v1/answers');
    const rawResponse = await output.json();
    const statsWithoutAverage = this.cleanStatistics(rawResponse);
    const response = this.getAverages(statsWithoutAverage);
    await this.setState({ response })
  }

  cleanStatistics = (response) => {
    return response.reduce( (answers, questionObj, index) => {
      if (!answers[questionObj.question]) {
        answers[questionObj.question] = { count: {}, sum: 0 };
      }

      //handle count
      if (!answers[questionObj.question].count[questionObj.user_answer]) {
        answers[questionObj.question].count[questionObj.user_answer] = 0;
      }

      answers[questionObj.question].count[questionObj.user_answer]++;

      //handle sum
      if (questionObj.user_answer.split('_').length === 1) {
        answers[questionObj.question].sum += parseInt(questionObj.user_answer, 10);
      } else {
        const splitArray = questionObj.user_answer.split('_')
        let number = parseInt(splitArray[0], 10);
        splitArray[1] === 'million' ? number *= 1000000 : number *= 1000000000;
        answers[questionObj.question].sum += number;
      }

      return answers
    }, {})
  }

  getAverages = (response) => {
    Object.keys(response).forEach(key => {
      const counts = Object.values(response[key].count);
      const totalCount = counts.reduce((totalCount, number) => {return totalCount += number}, 0);
      response[key].average = response[key].sum / totalCount
    });
    return response
  }

  displayStatistics = () => {
    const response = this.state.response;
    if (response) {
      return Object.keys(response).map( (key, index) => {
        if (response[key].average > 1000000) {
          const monetary = (response[key].average / 1000000000) > 0 ? 'billion' : 'million'
          const number = [...response[key].average.toString()];
          const fullNumber = `${number[0]}${number[1]}${number[2]} ${monetary}`
          return (
            <div className='answer'>
              <p>{ fullNumber }</p>
            </div>
          )
        }
      return (
        <div key={index} className='statistics-card'>
          <h4>{key}</h4>
            <hr/>
          <h6>Average Answer: {response[key].average}</h6>
        </div>
      )
      })
    }
  }

  render = () => {
    return (
      <div className='Statistics'>
        <h2>Whats the buzz on how people are answering?</h2>
        { this.displayStatistics() }
      </div>
    )
  }
}

export default Statistics;
