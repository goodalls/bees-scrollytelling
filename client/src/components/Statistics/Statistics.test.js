/* eslint-disable */
import React from 'react';
import Statistics from './Statistics';
import { shallow } from 'enzyme';
import mockData from '../../__mocks__/mockData';

describe('Statistics', () => {
  let wrapper;

  global.window.fetch = jest.fn().mockImplementation(() => {
    return Promise.resolve({
      status: 200,
      json: () => Promise.resolve([])
    });
  });

  beforeEach(() => {
    wrapper = shallow(<Statistics currentUserId={2} />)
  });

  it('matches the Snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('has a default state of an empty object', () => {
    wrapper = shallow(<Statistics currentUserId={2} />, { disableLifecycleMethods: true })
    expect(wrapper.instance().state).toEqual({});
  });

  describe('getStatistics', () => {
    it('gets the statistics and updates state with compiled data', async () => {
      wrapper.instance().cleanStatistics = jest.fn();
      wrapper.instance().getAverages = () => 'response';

      await wrapper.instance().getStatistics();

      expect(wrapper.instance().state.response).toEqual('response')
    });
  });

  describe('cleanStatistics', () => {

    it('returns a cleaned statistics data object', () => {
      expect(wrapper.instance().cleanStatistics(mockData.rawStatistics)).toEqual(mockData.cleanedStatistics);
    });
  });

    it('returns a cleaned statistics data object with averages', () => {
      expect(wrapper.instance().getAverages(mockData.cleanedStatistics)).toEqual(mockData.cleanedAverageData);
    });
});