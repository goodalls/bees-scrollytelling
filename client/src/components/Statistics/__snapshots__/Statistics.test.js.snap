// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Statistics matches the Snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Statistics
    currentUserId={2}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2>
          Whats the buzz on how people are answering?
</h2>,
        undefined,
      ],
      "className": "Statistics",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Whats the buzz on how people are answering?",
        },
        "ref": null,
        "rendered": "Whats the buzz on how people are answering?",
        "type": "h2",
      },
      undefined,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
            Whats the buzz on how people are answering?
</h2>,
          undefined,
        ],
        "className": "Statistics",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Whats the buzz on how people are answering?",
          },
          "ref": null,
          "rendered": "Whats the buzz on how people are answering?",
          "type": "h2",
        },
        undefined,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
